File: Training/Training/ETL-Part-3-1.0.3/Scala/Runnable/Runnable-5-Stream.scala
// Databricks notebook source

spark.sql("set spark.databricks.delta.preview.enabled=true") // Confirm Delta is enabled

// COMMAND ----------

// TODO

FILL_IN # create the "readPath" widget
FILL_IN # create the "writePath" widget
FILL_IN # create the "checkpointLocation" widget

// COMMAND ----------

// TODO

val readPath = FILL_IN            // Assign "readPath" to the value of the corresponding widget
val writePath = FILL_IN           // Assign "writePath" to the value of the corresponding widget
val checkpointLocation = FILL_IN  // Assign "checkpointLocation" to the value of the corresponding widget

// COMMAND ----------

try {
  // Make sure the readPath actually exists
  dbutils.fs.ls(readPath)
} catch {
  case e:Exception => dbutils.notebook.exit("Failed to find readPath")
}

// COMMAND ----------

// TODO 

Using a DataStreamReader, create a DataFrame with the previously
specified parameter: readPath

tweetsDF = FILL_IN

// COMMAND ----------

// TODO 
 
Using a DataStreamWriter, write the data to either 
a delta table or parquet file with the specified 
parameters: checkpointLocation, writePath & Trigger.Once()

tweetsDF.writeStream.FILL_IN 

// COMMAND ----------

val count = spark.read.format("delta").load(writePath).count()

// COMMAND ----------

// TODO

Exit the notebook returning the 
count from the previous cell

dbutils.FILL_IN

File: Training/Training/ETL-Part-3-1.0.3/Python/Solutions/Runnable/Runnable-5-Stream.py
# Databricks notebook source

spark.sql("set spark.databricks.delta.preview.enabled=true") # Confirm Delta is enabled

# COMMAND ----------

# ANSWER

dbutils.widgets.text("readPath", "")
dbutils.widgets.text("writePath", "")
dbutils.widgets.text("checkpointLocation", "")

# COMMAND ----------

# ANSWER

readPath = dbutils.widgets.get("readPath")
writePath = dbutils.widgets.get("writePath")
checkpointLocation = dbutils.widgets.get("checkpointLocation")

# COMMAND ----------

try:
  # Make sure the readPath actually exists
  dbutils.fs.ls(readPath)
except:
  dbutils.notebook.exit("Failed to find readPath")

# COMMAND ----------

# ANSWER

ddlSchema = "`tweet_id` BIGINT,`text` STRING,`screen_name` STRING,`user_id` BIGINT,`lang` STRING"

tweetsDF = (spark
  .readStream
  .option("path", readPath)
  .format("json")
  .schema(ddlSchema)
  .load()
)

# COMMAND ----------

# ANSWER

sq = (tweetsDF
  .writeStream
  .trigger(once=True)
  .format("delta") # This can be parquet if Delta is not enabled
  .option("checkpointLocation", checkpointLocation)
  .option("path", writePath)
  .outputMode("append")
  .start()
)
sq.awaitTermination()

# COMMAND ----------

try:
  dbutils.fs.ls(writePath)
except:
  dbutils.notebook.exit("Failed to find writePath")

# COMMAND ----------

count = spark.read.format("delta").load(writePath).count()

# COMMAND ----------

# ANSWER
dbutils.notebook.exit(str(count))
